File: Python/61.py
N, M = map(int,input().split())
INF = 10**9
box = [[INF] * N for _ in range(N)]
for i in range(M):
    a,b,t = map(int,input().split())
    a -= 1
    b -= 1
    box[a][b] = t
    box[b][a] = t
for i in range(N):
    box[i][i] = 0
for k in range(N):
    for i in range(N):
        for j in range(N):
            box[i][j] = min(box[i][j], box[i][k] + box[k][j])

ans = 10**9
for i in range(N):
    ans = min(ans, max(*box[i]))
print(ans)

File: Redcoder/C++/61.cpp
#include <bits/stdc++.h>
#define rep(i,nn,n) for (int i = nn; i < (n); ++i)
#define rrep(i,nn,n) for (int i = nn; i > (n); --i)
using namespace std;
using ll = long long;
using P = pair<int,int>;
#define chmin(x,y) x = min(x,y)
#define chmax(x,y) x = max(x,y)

const int INF = 1001001001;

int main() {

    int N,M;
    cin >> N >> M;
    vector<vector<int>> box(N, vector<int>(N,INF));
    rep(i,0,M){
        int a,b,t;
        cin >> a >> b >> t;
        --a; --b;
        box[a][b] = t;
        box[b][a] = t;
    }
    rep(i,0,N){
        box[i][i] = 0;
    }
    rep(k,0,N){
        rep(i,0,N){
            rep(j,0,N){
                box[i][j] = min(box[i][j], box[i][k] + box[k][j]);
            }
        }
    }
    int ans = INF;
    rep(i,0,N){
        int maxnum = 0;
        rep(j,0,N){
            chmax(maxnum, box[i][j]);
        }
        chmin(ans, maxnum);
    }
    cout << ans << endl;
    return 0;
}